### DISCUSSION TITLE
서버에 엑세스 토큰을 저장해야 할까요?

### DISCUSSION THREAD

**Discussion by minju412** [petit-a-petit/MIML#25] [General]
## 일반적인 경우
OAuth 2.0 프로토콜에서는 일반적으로 "클라이언트"에서 (Provider가 제공한) Access Token을 저장한다고 합니다. 
이때 저장한 Access Token은 클라이언트가 사용자의 음악 데이터에 엑세스하는 데 사용됩니다.

Access Token을 안전하게 저장하는 것은 보안적으로 중요한데, 클라이언트에서 저장하는 방법은 아래와 같습니다.

방법 1. OAuth 라이브러리 사용
OAuth 라이브러리를 사용하면 엑세스 토큰 저장 및 관리를 쉽게 할 수 있습니다. 대부분의 라이브러리는 보안을 고려하여 엑세스 토큰을 저장합니다.

방법 2. 세션관리
엑세스 토큰을 사용할 때마다 임시로 저장하고, 앱이 종료될 때나 사용자가 로그아웃할 때 해당 토큰을 삭제합니다. 이 방법은 엑세스 토큰을 최소한의 시간 동안만 저장하고 사용합니다.

⚠️ 서버에서 사용자의 엑세스 토큰을 저장하는 것은 일반적이지 않다고 합니다. 그 이유는 클라이언트 애플리케이션에서 엑세스 토큰이 사용되기 때문입니다.

OAuth 2.0 프로토콜에서 서버의 역할을 아래와 같습니다.
1. 클라이언트에게 엑세스 토큰 제공: Provider로부터 얻은 엑세스 토큰을 클라이언트에게 전달합니다.
2. 클라이언트 인증: Provider에 대한 클라이언트 자격 증명을 관리합니다.

## 우리 앱의 경우
- 상황: 클라이언트는 구현하지 않지만, Spotify 내 음악 데이터에는 접근해야 한다.
- 해결: Spring Security 기능 활용

```java
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.core.user.OAuth2User;

@RestController
public class UserController {

    @GetMapping("/user")
    public String getUserInfo(@AuthenticationPrincipal OAuth2User principal) {
        String accessToken = principal.getAttribute("access_token");
        
        // 엑세스 토큰을 저장하거나 필요한 작업 수행
        // 예: 데이터베이스에 저장
        // 예: 엑세스 토큰을 사용하여 Spotify API 호출

        return "User Info: " + principal.getName();
    }
}
```

위 코드에서 `@AuthenticationPrincipal OAuth2User principal`은 현재 사용자 정보를 제공합니다. `getAttribute("access_token")` 메서드를 사용하여 Spotify에서 제공한 엑세스 토큰을 가져올 수 있습니다. 
필요에 따라 엑세스 토큰을 데이터베이스에 저장하거나 사용할 수 있습니다.

데이터베이스에 저장한다면 이전 회의 시간에 이야기 했던 것과 같이 redis를 사용하여 엑세스 토큰을 저장하고, expire를 지정하는 방법이 있습니다.

**하지만 Spotify API를 사용하기 위한 목적이라면 굳이 저장하지 않아도 될 것 같은데 다양한 의견을 듣고 싶습니다 😊**

**Comment by Jeonwoojae** [comment]
네 스포티파이의 음악과 사용자 정보 API만을 사용하기 위해서는 스포티파이의 액세스 토큰을 DB에 저장하지 않아도 될 것 같습니다. 

하지만 친구를 신청하고 친구 관계를 접근하기 위해서는 현재 사용자가 해당 기능에 권한이 있는지 확인할 필요가 있을 것 같은데 해당 권한에 대한 정보는 어떻게 관리하시나요?? principal에 현재 유저 정보를 넣어두시나요?

**Reply by minju412** [reply]
@Jeonwoojae 앗 생각해보니 친구 신청 같이 스포티파이 API를 사용하지 않고 저희가 직접 구현한 로직에 대한 접근 권한을 확인하려면, 스포티파이에서 내려준 엑세스 토큰 뿐만 아니라 저희 앱 고유의 엑세스 토큰 발급이 불가피 할 것 같은데 어떻게 생각하시나요?!

**Reply by Jeonwoojae** [reply]
네 내부 권한 인증을 토큰으로 하신다면 토큰 발급이 불가피 할 것 같습니다!
